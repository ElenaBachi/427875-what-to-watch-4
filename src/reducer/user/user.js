import {extend} from "../../utils/utils.js";
import {isValidEmail, isValidPassword} from "../../utils/validation.js";
import {ErrorMessage} from "../../consts.js";

const AuthorizationStatus = {
  AUTH: `AUTH`,
  NO_AUTH: `NO_AUTH`,
};

const initialState = {
  authorizationStatus: AuthorizationStatus.NO_AUTH,
  authorizationErrorMessage: ``,
};

const ActionType = {
  REQUIRED_AUTHORIZATION: `REQUIRED_AUTHORIZATION`,
  SET_AUTHORIZATION_ERROR_MESSAGE: `SET_AUTHORIZATION_ERROR_MESSAGE`,
};

const ActionCreator = {
  requireAuthorization: (status) => {
    return {
      type: ActionType.REQUIRED_AUTHORIZATION,
      payload: status,
    };
  },

  setAuthorizationErrorMessage: (errorMessage) => {
    return {
      type: ActionType.SET_AUTHORIZATION_ERROR_MESSAGE,
      payload: errorMessage,
    };
  }
};

const Operation = {
  checkAuth: () => (dispatch, getState, api) => {
    return api.get(`login`)
      .then(() => {
        dispatch(ActionCreator.requireAuthorization(AuthorizationStatus.AUTH));
      })
      .catch((err) => {
        throw err;
      });
  },

  login: (authData) => (dispatch, getState, api) => {
    const validEmail = isValidEmail(authData.login);
    const validPassword = isValidPassword(authData.password);

    if (!validEmail) {
      dispatch(ActionCreator.setAuthorizationErrorMessage(ErrorMessage.LOGIN));
    } else if (!validPassword) {
      dispatch(ActionCreator.setAuthorizationErrorMessage(ErrorMessage.PASSWORD));
    }

    return api.post(`login`, {
      email: authData.login,
      password: authData.password,
    })
      .then(() => {
        dispatch(ActionCreator.requireAuthorization(AuthorizationStatus.AUTH));
        dispatch(ActionCreator.setAuthorizationErrorMessage(``));
      })
      .catch((err) => {
        if (err.status === 401) {
          dispatch(ActionCreator.setAuthorizationErrorMessage(ErrorMessage.UNRECOGNIZED_DATA));
        } else if (err.status === 400) {
          dispatch(ActionCreator.setAuthorizationErrorMessage(ErrorMessage.INCOMPLETE_DATA));
        }

        dispatch(ActionCreator.setAuthorizationErrorMessage(err.data.error));
      });
  },
};

const reducer = (state = initialState, action) => {
  switch (action.type) {
    case ActionType.REQUIRED_AUTHORIZATION:
      return extend(state, {
        authorizationStatus: action.payload,
      });
    case ActionType.SET_AUTHORIZATION_ERROR_MESSAGE:
      return extend(state, {
        authorizationErrorMessage: action.payload,
      });
  }

  return state;
};

export {reducer, Operation, ActionType, ActionCreator, AuthorizationStatus};
