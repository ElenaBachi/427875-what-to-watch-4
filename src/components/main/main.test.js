import React from "react";
import renderer from "react-test-renderer";
import {Provider} from 'react-redux';
import configureStore from 'redux-mock-store';

import Main from "./main.jsx";

import NameSpace from "../../reducer/name-space.js";
import {AuthorizationStatus} from "../../reducer/user/user.js";

const mockStore = configureStore([]);

const mock = {
  films: [
    {
      id: 1,
      title: `title1`,
      genre: `genre1`,
      year: 2020,
      img: `img-1.jpg`,
      poster: `poster-1.jpg`,
      cover: `cover-1.jpg`,
      videoSrc: `video-link-1`,
      previewVideoSrc: `preview-video-link-1`,
      description: `description1`,
      score: 9,
      count: 200,
      director: `director1`,
      actorList: [`Actor1`, `Actor2`, `Actor3`],
      runTime: 100,
      isFavorite: false,
      bgColor: `fff`,
    }, {
      id: 2,
      title: `title2`,
      genre: `genre2`,
      year: 2019,
      img: `img-2.jpg`,
      poster: `poster-2.jpg`,
      cover: `cover-2.jpg`,
      videoSrc: `video-link-2`,
      previewVideoSrc: `preview-video-link-2`,
      description: `description2`,
      score: 9,
      count: 200,
      director: `director2`,
      actorList: [`Actor1`, `Actor2`, `Actor3`],
      runTime: 100,
      isFavorite: false,
      bgColor: `fff`,
    }, {
      id: 3,
      title: `title3`,
      genre: `genre3`,
      year: 2018,
      img: `img-3.jpg`,
      poster: `poster-3.jpg`,
      cover: `cover-3.jpg`,
      videoSrc: `video-link-3`,
      previewVideoSrc: `preview-video-link-3`,
      description: `description3`,
      score: 9,
      count: 200,
      director: `director3`,
      actorList: [`Actor1`, `Actor2`, `Actor3`],
      runTime: 100,
      isFavorite: false,
      bgColor: `fff`,
    },
  ],
  promoFilm: {
    id: 99,
    title: `title`,
    genre: `genre`,
    year: 2020,
    img: `img.jpg`,
    poster: `poster.jpg`,
    cover: `cover.jpg`,
    videoSrc: `video-link`,
    previewVideoSrc: `preview-video-link`,
    description: `description`,
    score: 9,
    count: 200,
    director: `director`,
    actorList: [`Actor1`, `Actor2`, `Actor3`],
    runTime: 100,
  },
};

describe(`Render Main component`, () => {
  it(`Should render Main correctly`, () => {
    const store = mockStore({
      [NameSpace.DATA]: {
        films: mock.films,
        promoFilm: mock.promoFilm,
      },
      [NameSpace.FILTER]: {
        currentFilter: `All genres`,
      },
      [NameSpace.FILMS_LOAD_BTN]: {
        filmCount: 10,
      },
      [NameSpace.USER]: {
        authorizationStatus: AuthorizationStatus.NO_AUTH,
        authorizationErrorMessage: ``,
      },
    });

    const onShowMoreBtnClick = () => {};
    const handlePlayButtonClick = () => {};
    const onScreenChange = () => {};

    const tree = renderer.create(
        <Provider store={store}>
          <Main
            onShowMoreBtnClick={onShowMoreBtnClick}
            handlePlayButtonClick={handlePlayButtonClick}
            onScreenChange={onScreenChange}
          />
        </Provider>, {
          createNodeMock: () => {
            return {};
          }
        }).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
