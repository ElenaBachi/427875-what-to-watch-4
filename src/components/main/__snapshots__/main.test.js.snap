// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render Main component Should render Main correctly 1`] = `
Array [
  <section
    className="movie-card"
  >
    <div
      className="movie-card__bg"
    >
      <img
        alt="title"
        src="cover.jpg"
      />
    </div>
    <h1
      className="visually-hidden"
    >
      WTW
    </h1>
    <header
      className="page-header movie-card__head"
    >
      <div
        className="logo"
      >
        <a
          className="logo__link"
          href="/"
          onClick={[Function]}
        >
          <span
            className="logo__letter logo__letter--1"
          >
            W
          </span>
          <span
            className="logo__letter logo__letter--2"
          >
            T
          </span>
          <span
            className="logo__letter logo__letter--3"
          >
            W
          </span>
        </a>
      </div>
      <div
        className="user-block"
      >
        <a
          className="user-block__link"
          href="/login"
          onClick={[Function]}
        >
          Sign in
        </a>
      </div>
    </header>
    <div
      className="movie-card__wrap"
    >
      <div
        className="movie-card__info"
      >
        <div
          className="movie-card__poster"
        >
          <img
            alt="title"
            height="327"
            src="poster.jpg"
            width="218"
          />
        </div>
        <div
          className="movie-card__desc"
        >
          <h2
            className="movie-card__title"
          >
            title
          </h2>
          <p
            className="movie-card__meta"
          >
            <span
              className="movie-card__genre"
            >
              genre
            </span>
            <span
              className="movie-card__year"
            >
              2020
            </span>
          </p>
          <div
            className="movie-card__buttons"
          >
            <button
              className="btn btn--play movie-card__button"
              onClick={[Function]}
              type="button"
            >
              <svg
                height="19"
                viewBox="0 0 19 19"
                width="19"
              >
                <use
                  xlinkHref="#play-s"
                />
              </svg>
              <span>
                Play
              </span>
            </button>
            <button
              className="btn btn--list movie-card__button"
              onClick={[Function]}
              type="button"
            >
              <svg
                height="20"
                viewBox="0 0 19 20"
                width="19"
              >
                <use
                  xlinkHref="#add"
                />
              </svg>
              <span>
                My list
              </span>
            </button>
          </div>
        </div>
      </div>
    </div>
  </section>,
  <div
    className="page-content"
  >
    <section
      className="catalog"
    >
      <h2
        className="catalog__title visually-hidden"
      >
        Catalog
      </h2>
      <ul
        className="catalog__genres-list"
      >
        <li
          className="catalog__genres-item catalog__genres-item--active"
          onClick={[Function]}
        >
          <a
            className="catalog__genres-link"
            href="#"
          >
            All genres
          </a>
        </li>
        <li
          className="catalog__genres-item "
          onClick={[Function]}
        >
          <a
            className="catalog__genres-link"
            href="#"
          >
            genre1
          </a>
        </li>
        <li
          className="catalog__genres-item "
          onClick={[Function]}
        >
          <a
            className="catalog__genres-link"
            href="#"
          >
            genre2
          </a>
        </li>
        <li
          className="catalog__genres-item "
          onClick={[Function]}
        >
          <a
            className="catalog__genres-link"
            href="#"
          >
            genre3
          </a>
        </li>
      </ul>
      <div
        className="catalog__movies-list"
      >
        <article
          className="small-movie-card catalog__movies-card"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          <div
            className="small-movie-card__image"
          >
            <video
              autoPlay={false}
              height="175"
              loop={true}
              poster="img-1.jpg"
              width="280"
            />
          </div>
          <h3
            className="small-movie-card__title"
          >
            <a
              className="small-movie-card__link"
            >
              title1
            </a>
          </h3>
        </article>
        <article
          className="small-movie-card catalog__movies-card"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          <div
            className="small-movie-card__image"
          >
            <video
              autoPlay={false}
              height="175"
              loop={true}
              poster="img-2.jpg"
              width="280"
            />
          </div>
          <h3
            className="small-movie-card__title"
          >
            <a
              className="small-movie-card__link"
            >
              title2
            </a>
          </h3>
        </article>
        <article
          className="small-movie-card catalog__movies-card"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          <div
            className="small-movie-card__image"
          >
            <video
              autoPlay={false}
              height="175"
              loop={true}
              poster="img-3.jpg"
              width="280"
            />
          </div>
          <h3
            className="small-movie-card__title"
          >
            <a
              className="small-movie-card__link"
            >
              title3
            </a>
          </h3>
        </article>
      </div>
      <div
        className="catalog__more"
      />
    </section>
    <footer
      className="page-footer"
    >
      <div
        className="logo"
      >
        <a
          className="logo__link logo__link--light"
        >
          <span
            className="logo__letter logo__letter--1"
          >
            W
          </span>
          <span
            className="logo__letter logo__letter--2"
          >
            T
          </span>
          <span
            className="logo__letter logo__letter--3"
          >
            W
          </span>
        </a>
      </div>
      <div
        className="copyright"
      >
        <p>
          Â© 2019 What to watch Ltd.
        </p>
      </div>
    </footer>
  </div>,
]
`;
